File: hdf-diff/go/matching/srg_deterministic_test.go
package matching

import (
	"testing"

	hdf "github.com/mitre/hdf-libs/hdf-schema/dist/go/v3"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// helper to build a requirement with gtitle tag.
func reqWithGtitle(id, gtitle string) hdf.EvaluatedRequirement {
	return hdf.EvaluatedRequirement{
		ID:     id,
		Impact: 0.7,
		Tags:   map[string]any{"gtitle": gtitle},
	}
}

func TestSrgDeterministicStrategy_Name(t *testing.T) {
	s := NewSRGDeterministicStrategy()
	assert.Equal(t, "srgDeterministic", s.Name())
}

func TestSrgDeterministicStrategy_Match1to1(t *testing.T) {
	s := NewSRGDeterministicStrategy()
	oldReqs := []hdf.EvaluatedRequirement{reqWithGtitle("V-001", "SRG-OS-000185-GPOS-00079")}
	newReqs := []hdf.EvaluatedRequirement{reqWithGtitle("RHEL-001", "SRG-OS-000185-GPOS-00079")}

	result := s.Match(oldReqs, newReqs)

	require.Len(t, result.Matched, 1)
	assert.Equal(t, "V-001", result.Matched[0].OldReq.ID)
	assert.Equal(t, "RHEL-001", result.Matched[0].NewReq.ID)
	assert.Equal(t, "srgDeterministic", result.Matched[0].Strategy)
	assert.Equal(t, 1.0, result.Matched[0].Confidence)
	assert.Equal(t, "primary", result.Matched[0].Relationship)
	assert.Len(t, result.UnmatchedOld, 0)
	assert.Len(t, result.UnmatchedNew, 0)
}

func TestSrgDeterministicStrategy_NNewOneOld(t *testing.T) {
	s := NewSRGDeterministicStrategy()
	oldReqs := []hdf.EvaluatedRequirement{reqWithGtitle("V-001", "SRG-OS-000185-GPOS-00079")}
	newReqs := []hdf.EvaluatedRequirement{
		reqWithGtitle("RHEL-001", "SRG-OS-000185-GPOS-00079"),
		reqWithGtitle("RHEL-002", "SRG-OS-000185-GPOS-00079"),
		reqWithGtitle("RHEL-003", "SRG-OS-000185-GPOS-00079"),
	}

	result := s.Match(oldReqs, newReqs)

	require.Len(t, result.Matched, 3)
	assert.Len(t, result.UnmatchedOld, 0)
	assert.Len(t, result.UnmatchedNew, 0)

	// First new gets primary
	assert.Equal(t, "RHEL-001", result.Matched[0].NewReq.ID)
	assert.Equal(t, "primary", result.Matched[0].Relationship)

	// Rest get related
	assert.Equal(t, "RHEL-002", result.Matched[1].NewReq.ID)
	assert.Equal(t, "related", result.Matched[1].Relationship)
	assert.Equal(t, "RHEL-003", result.Matched[2].NewReq.ID)
	assert.Equal(t, "related", result.Matched[2].Relationship)

	// All reference same old req
	for _, m := range result.Matched {
		assert.Equal(t, "V-001", m.OldReq.ID)
		assert.Equal(t, 1.0, m.Confidence)
	}
}

func TestSrgDeterministicStrategy_OneNewNOld(t *testing.T) {
	s := NewSRGDeterministicStrategy()
	oldReqs := []hdf.EvaluatedRequirement{
		reqWithGtitle("V-001", "SRG-OS-000185-GPOS-00079"),
		reqWithGtitle("V-002", "SRG-OS-000185-GPOS-00079"),
	}
	newReqs := []hdf.EvaluatedRequirement{reqWithGtitle("RHEL-001", "SRG-OS-000185-GPOS-00079")}

	result := s.Match(oldReqs, newReqs)

	require.Len(t, result.Matched, 2)
	assert.Len(t, result.UnmatchedOld, 0)
	assert.Len(t, result.UnmatchedNew, 0)

	// First old gets primary
	assert.Equal(t, "V-001", result.Matched[0].OldReq.ID)
	assert.Equal(t, "primary", result.Matched[0].Relationship)

	// Second old gets related
	assert.Equal(t, "V-002", result.Matched[1].OldReq.ID)
	assert.Equal(t, "related", result.Matched[1].Relationship)

	// Both reference same new req
	for _, m := range result.Matched {
		assert.Equal(t, "RHEL-001", m.NewReq.ID)
	}
}

func TestSrgDeterministicStrategy_NMAmbiguous(t *testing.T) {
	s := NewSRGDeterministicStrategy()
	oldReqs := []hdf.EvaluatedRequirement{
		reqWithGtitle("V-001", "SRG-OS-000185-GPOS-00079"),
		reqWithGtitle("V-002", "SRG-OS-000185-GPOS-00079"),
	}
	newReqs := []hdf.EvaluatedRequirement{
		reqWithGtitle("RHEL-001", "SRG-OS-000185-GPOS-00079"),
		reqWithGtitle("RHEL-002", "SRG-OS-000185-GPOS-00079"),
	}

	result := s.Match(oldReqs, newReqs)

	assert.Len(t, result.Matched, 0)
	assert.Len(t, result.UnmatchedOld, 2)
	assert.Len(t, result.UnmatchedNew, 2)
}

func TestSrgDeterministicStrategy_NoGtitle(t *testing.T) {
	s := NewSRGDeterministicStrategy()
	oldReqs := []hdf.EvaluatedRequirement{{ID: "V-001", Impact: 0.7, Tags: map[string]any{"nist": "AC-1"}}}
	newReqs := []hdf.EvaluatedRequirement{{ID: "RHEL-001", Impact: 0.7, Tags: map[string]any{"nist": "AC-1"}}}

	result := s.Match(oldReqs, newReqs)

	assert.Len(t, result.Matched, 0)
	assert.Len(t, result.UnmatchedOld, 1)
	assert.Len(t, result.UnmatchedNew, 1)
}

func TestSrgDeterministicStrategy_NoTags(t *testing.T) {
	s := NewSRGDeterministicStrategy()
	oldReqs := []hdf.EvaluatedRequirement{{ID: "V-001", Impact: 0.7}}
	newReqs := []hdf.EvaluatedRequirement{{ID: "RHEL-001", Impact: 0.7}}

	result := s.Match(oldReqs, newReqs)

	assert.Len(t, result.Matched, 0)
	assert.Len(t, result.UnmatchedOld, 1)
	assert.Len(t, result.UnmatchedNew, 1)
}

func TestSrgDeterministicStrategy_MixedMatchAndNoMatch(t *testing.T) {
	s := NewSRGDeterministicStrategy()
	oldReqs := []hdf.EvaluatedRequirement{
		reqWithGtitle("V-001", "SRG-OS-000001"),
		reqWithGtitle("V-002", "SRG-OS-000002"),
		{ID: "V-003", Impact: 0.3},
	}
	newReqs := []hdf.EvaluatedRequirement{
		reqWithGtitle("RHEL-001", "SRG-OS-000001"),
		reqWithGtitle("RHEL-002", "SRG-OS-000999"),
		{ID: "RHEL-003", Impact: 0.3},
	}

	result := s.Match(oldReqs, newReqs)

	require.Len(t, result.Matched, 1)
	assert.Equal(t, "V-001", result.Matched[0].OldReq.ID)
	assert.Equal(t, "RHEL-001", result.Matched[0].NewReq.ID)
	assert.Equal(t, "primary", result.Matched[0].Relationship)

	assert.Len(t, result.UnmatchedOld, 2)
	assert.Len(t, result.UnmatchedNew, 2)
}

func TestSrgDeterministicStrategy_MultipleDistinctGtitles(t *testing.T) {
	s := NewSRGDeterministicStrategy()
	oldReqs := []hdf.EvaluatedRequirement{
		reqWithGtitle("V-001", "SRG-OS-000001"),
		reqWithGtitle("V-002", "SRG-OS-000002"),
	}
	newReqs := []hdf.EvaluatedRequirement{
		reqWithGtitle("RHEL-001", "SRG-OS-000001"),
		reqWithGtitle("RHEL-002", "SRG-OS-000002"),
	}

	result := s.Match(oldReqs, newReqs)

	require.Len(t, result.Matched, 2)
	// Map iteration order not guaranteed — check by content not position
	matchedPairs := make(map[string]string)
	for _, m := range result.Matched {
		matchedPairs[m.OldReq.ID] = m.NewReq.ID
	}
	assert.Equal(t, "RHEL-001", matchedPairs["V-001"])
	assert.Equal(t, "RHEL-002", matchedPairs["V-002"])
	assert.Len(t, result.UnmatchedOld, 0)
	assert.Len(t, result.UnmatchedNew, 0)
}

func TestSrgDeterministicStrategy_EmptyInputs(t *testing.T) {
	s := NewSRGDeterministicStrategy()
	result := s.Match(nil, nil)
	assert.Len(t, result.Matched, 0)
	assert.Len(t, result.UnmatchedOld, 0)
	assert.Len(t, result.UnmatchedNew, 0)
}

func TestSrgDeterministicStrategy_EmptyGtitle(t *testing.T) {
	s := NewSRGDeterministicStrategy()
	oldReqs := []hdf.EvaluatedRequirement{
		{ID: "V-001", Impact: 0.7, Tags: map[string]any{"gtitle": ""}},
	}
	newReqs := []hdf.EvaluatedRequirement{
		{ID: "RHEL-001", Impact: 0.7, Tags: map[string]any{"gtitle": ""}},
	}

	result := s.Match(oldReqs, newReqs)

	assert.Len(t, result.Matched, 0)
	assert.Len(t, result.UnmatchedOld, 1)
	assert.Len(t, result.UnmatchedNew, 1)
}

func TestSrgDeterministicStrategy_GtitleNotString(t *testing.T) {
	s := NewSRGDeterministicStrategy()
	oldReqs := []hdf.EvaluatedRequirement{
		{ID: "V-001", Impact: 0.7, Tags: map[string]any{"gtitle": 42}},
	}
	newReqs := []hdf.EvaluatedRequirement{
		{ID: "RHEL-001", Impact: 0.7, Tags: map[string]any{"gtitle": 42}},
	}

	result := s.Match(oldReqs, newReqs)

	assert.Len(t, result.Matched, 0)
	assert.Len(t, result.UnmatchedOld, 1)
	assert.Len(t, result.UnmatchedNew, 1)
}

File: hdf-diff/test/matching/srg-deterministic.test.ts
import { describe, it, expect } from 'vitest';
import { createSrgDeterministicStrategy } from '../../src/matching/srg-deterministic.js';

/** Helper to create a requirement with gtitle tag. */
function reqWithGtitle(id: string, gtitle: string, title?: string) {
  return {
    id,
    impact: 0.7,
    title: title ?? `Control ${id}`,
    tags: { gtitle },
  };
}

describe('SrgDeterministicStrategy', () => {
  const strategy = createSrgDeterministicStrategy();

  it('should have name "srgDeterministic"', () => {
    expect(strategy.name).toBe('srgDeterministic');
  });

  it('should match 1:1 by gtitle with confidence 1.0 and relationship "primary"', () => {
    const oldReqs = [reqWithGtitle('V-001', 'SRG-OS-000185-GPOS-00079')];
    const newReqs = [reqWithGtitle('RHEL-001', 'SRG-OS-000185-GPOS-00079')];

    const result = strategy.match(oldReqs, newReqs);

    expect(result.matched).toHaveLength(1);
    expect(result.matched[0]!.oldReq['id']).toBe('V-001');
    expect(result.matched[0]!.newReq['id']).toBe('RHEL-001');
    expect(result.matched[0]!.strategy).toBe('srgDeterministic');
    expect(result.matched[0]!.confidence).toBe(1.0);
    expect(result.matched[0]!.relationship).toBe('primary');
    expect(result.unmatchedOld).toHaveLength(0);
    expect(result.unmatchedNew).toHaveLength(0);
  });

  it('should handle N new + 1 old (1:N split) — first primary, rest related', () => {
    const oldReqs = [reqWithGtitle('V-001', 'SRG-OS-000185-GPOS-00079')];
    const newReqs = [
      reqWithGtitle('RHEL-001', 'SRG-OS-000185-GPOS-00079'),
      reqWithGtitle('RHEL-002', 'SRG-OS-000185-GPOS-00079'),
      reqWithGtitle('RHEL-003', 'SRG-OS-000185-GPOS-00079'),
    ];

    const result = strategy.match(oldReqs, newReqs);

    expect(result.matched).toHaveLength(3);
    expect(result.unmatchedOld).toHaveLength(0);
    expect(result.unmatchedNew).toHaveLength(0);

    // First new gets primary
    expect(result.matched[0]!.newReq['id']).toBe('RHEL-001');
    expect(result.matched[0]!.relationship).toBe('primary');

    // Rest get related
    expect(result.matched[1]!.newReq['id']).toBe('RHEL-002');
    expect(result.matched[1]!.relationship).toBe('related');
    expect(result.matched[2]!.newReq['id']).toBe('RHEL-003');
    expect(result.matched[2]!.relationship).toBe('related');

    // All reference the same old req
    for (const m of result.matched) {
      expect(m.oldReq['id']).toBe('V-001');
      expect(m.confidence).toBe(1.0);
    }
  });

  it('should handle 1 new + N old — first old primary, rest related', () => {
    const oldReqs = [
      reqWithGtitle('V-001', 'SRG-OS-000185-GPOS-00079'),
      reqWithGtitle('V-002', 'SRG-OS-000185-GPOS-00079'),
    ];
    const newReqs = [reqWithGtitle('RHEL-001', 'SRG-OS-000185-GPOS-00079')];

    const result = strategy.match(oldReqs, newReqs);

    expect(result.matched).toHaveLength(2);
    expect(result.unmatchedOld).toHaveLength(0);
    expect(result.unmatchedNew).toHaveLength(0);

    // First old gets primary
    expect(result.matched[0]!.oldReq['id']).toBe('V-001');
    expect(result.matched[0]!.relationship).toBe('primary');

    // Second old gets related
    expect(result.matched[1]!.oldReq['id']).toBe('V-002');
    expect(result.matched[1]!.relationship).toBe('related');

    // Both reference the same new req
    for (const m of result.matched) {
      expect(m.newReq['id']).toBe('RHEL-001');
    }
  });

  it('should skip N:M ambiguous (multiple old and multiple new)', () => {
    const oldReqs = [
      reqWithGtitle('V-001', 'SRG-OS-000185-GPOS-00079'),
      reqWithGtitle('V-002', 'SRG-OS-000185-GPOS-00079'),
    ];
    const newReqs = [
      reqWithGtitle('RHEL-001', 'SRG-OS-000185-GPOS-00079'),
      reqWithGtitle('RHEL-002', 'SRG-OS-000185-GPOS-00079'),
    ];

    const result = strategy.match(oldReqs, newReqs);

    expect(result.matched).toHaveLength(0);
    expect(result.unmatchedOld).toHaveLength(2);
    expect(result.unmatchedNew).toHaveLength(2);
  });

  it('should pass through requirements with no gtitle', () => {
    const oldReqs = [{ id: 'V-001', impact: 0.7, tags: { nist: ['AC-1'] } }];
    const newReqs = [{ id: 'RHEL-001', impact: 0.7, tags: { nist: ['AC-1'] } }];

    const result = strategy.match(oldReqs, newReqs);

    expect(result.matched).toHaveLength(0);
    expect(result.unmatchedOld).toHaveLength(1);
    expect(result.unmatchedNew).toHaveLength(1);
  });

  it('should pass through requirements with no tags', () => {
    const oldReqs = [{ id: 'V-001', impact: 0.7 }];
    const newReqs = [{ id: 'RHEL-001', impact: 0.7 }];

    const result = strategy.match(oldReqs, newReqs);

    expect(result.matched).toHaveLength(0);
    expect(result.unmatchedOld).toHaveLength(1);
    expect(result.unmatchedNew).toHaveLength(1);
  });

  it('should handle mixed: some gtitles match, some do not', () => {
    const oldReqs = [
      reqWithGtitle('V-001', 'SRG-OS-000001'),
      reqWithGtitle('V-002', 'SRG-OS-000002'),
      { id: 'V-003', impact: 0.3 }, // no gtitle
    ];
    const newReqs = [
      reqWithGtitle('RHEL-001', 'SRG-OS-000001'),
      reqWithGtitle('RHEL-002', 'SRG-OS-000999'), // no match
      { id: 'RHEL-003', impact: 0.3 }, // no gtitle
    ];

    const result = strategy.match(oldReqs, newReqs);

    expect(result.matched).toHaveLength(1);
    expect(result.matched[0]!.oldReq['id']).toBe('V-001');
    expect(result.matched[0]!.newReq['id']).toBe('RHEL-001');
    expect(result.matched[0]!.relationship).toBe('primary');

    // V-002 unmatched (no new has SRG-OS-000002), V-003 unmatched (no gtitle)
    expect(result.unmatchedOld).toHaveLength(2);
    // RHEL-002 unmatched (no old has SRG-OS-000999), RHEL-003 unmatched (no gtitle)
    expect(result.unmatchedNew).toHaveLength(2);
  });

  it('should handle multiple distinct gtitles independently', () => {
    const oldReqs = [
      reqWithGtitle('V-001', 'SRG-OS-000001'),
      reqWithGtitle('V-002', 'SRG-OS-000002'),
    ];
    const newReqs = [
      reqWithGtitle('RHEL-001', 'SRG-OS-000001'),
      reqWithGtitle('RHEL-002', 'SRG-OS-000002'),
    ];

    const result = strategy.match(oldReqs, newReqs);

    expect(result.matched).toHaveLength(2);
    expect(result.matched[0]!.oldReq['id']).toBe('V-001');
    expect(result.matched[0]!.newReq['id']).toBe('RHEL-001');
    expect(result.matched[1]!.oldReq['id']).toBe('V-002');
    expect(result.matched[1]!.newReq['id']).toBe('RHEL-002');
    expect(result.unmatchedOld).toHaveLength(0);
    expect(result.unmatchedNew).toHaveLength(0);
  });

  it('should handle empty inputs', () => {
    expect(strategy.match([], []).matched).toHaveLength(0);
    expect(strategy.match([], []).unmatchedOld).toHaveLength(0);
    expect(strategy.match([], []).unmatchedNew).toHaveLength(0);
  });

  it('should handle gtitle present only in old (no new match)', () => {
    const oldReqs = [reqWithGtitle('V-001', 'SRG-OS-000001')];
    const newReqs = [reqWithGtitle('RHEL-001', 'SRG-OS-000999')];

    const result = strategy.match(oldReqs, newReqs);

    expect(result.matched).toHaveLength(0);
    expect(result.unmatchedOld).toHaveLength(1);
    expect(result.unmatchedNew).toHaveLength(1);
  });

  it('should handle empty string gtitle as missing', () => {
    const oldReqs = [{ id: 'V-001', impact: 0.7, tags: { gtitle: '' } }];
    const newReqs = [{ id: 'RHEL-001', impact: 0.7, tags: { gtitle: '' } }];

    const result = strategy.match(oldReqs, newReqs);

    expect(result.matched).toHaveLength(0);
    expect(result.unmatchedOld).toHaveLength(1);
    expect(result.unmatchedNew).toHaveLength(1);
  });
});
